Bug report: Tests halt on Rocky Linux 8 using PySide6-6.7.1

## Environment

OS: _Linux_
Python version: _3.11_
Toolkit: _Qt_
Qt API: _PySide6_ 6.7.1

## Description

Tests halt on (WSL2) Rocky Linux 8 using PySide6-6.7.1. The test run will resume if the pop up windows are closed but similar errors appear later and the run does not finish.

## Steps to Reproduce
```
edm env create test --version 3.11
edm install -e test PySide6 numpy pygments packaging pillow importlib_resources build wheel
edm run -e test -- pip install traits pyface
edm run -e test -- python -m unittest discover pyface
```

output
```
test_set_title_change_active_task (pyface.tasks.tests.test_task_window.TestTaskWindow.test_set_title_change_active_task) ... ok
test_set_title_change_active_task_name (pyface.tasks.tests.test_task_window.TestTaskWindow.test_set_title_change_active_task_name) ... ok
test_set_title_change_title (pyface.tasks.tests.test_task_window.TestTaskWindow.test_set_title_change_title) ... ok
test_set_title_no_active_task (pyface.tasks.tests.test_task_window.TestTaskWindow.test_set_title_no_active_task) ... ok
test_set_title_no_tasks (pyface.tasks.tests.test_task_window.TestTaskWindow.test_set_title_no_tasks) ... ok
test_title_activate_task (pyface.tasks.tests.test_task_window.TestTaskWindow.test_title_activate_task) ... ok
test_title_change_active_task (pyface.tasks.tests.test_task_window.TestTaskWindow.test_title_change_active_task) ... ok
test_title_change_active_task_name (pyface.tasks.tests.test_task_window.TestTaskWindow.test_title_change_active_task_name) ... ok
test_title_change_deactivate_task (pyface.tasks.tests.test_task_window.TestTaskWindow.test_title_change_deactivate_task) ... ok
test_title_default (pyface.tasks.tests.test_task_window.TestTaskWindow.test_title_default) ... ok
test_title_no_active_task (pyface.tasks.tests.test_task_window.TestTaskWindow.test_title_no_active_task) ... ok
test_defaults (pyface.tasks.tests.test_tasks_application.TestApplication.test_defaults) ... ok
test_lifecycle (pyface.tasks.tests.test_tasks_application.TestApplication.test_lifecycle) ... Traceback (most recent call last):
  File "/home/itziakos/.edm/envs/build-env/lib/python3.11/site-packages/pyface/ui/qt/gui.py", line 190, in _dispatch
    self._callable(*self._args, **self._kw)
  File "/home/itziakos/.edm/envs/build-env/lib/python3.11/site-packages/pyface/application.py", line 220, in exit
    self._exit()
  File "/home/itziakos/.edm/envs/build-env/lib/python3.11/site-packages/pyface/gui_application.py", line 236, in _exit
    self.gui.stop_event_loop()
    ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: '_Undefined' object has no attribute 'stop_event_loop'
```

![image](https://github.com/enthought/pyface/assets/403930/21224ff7-925c-4d93-8fc1-07b34b23f240)

On WSL2 the tests halt but on linux VM they segfault.
```
[2024-06-12T15:33:10.219Z] b'test_lifecycle (pyface.tasks.tests.test_tasks_application.TestApplication.test_lifecycle) ... Traceback (most recent call last):'
[2024-06-12T15:33:10.220Z] b'  File "/home/builder/.edm/envs/build-env/lib/python3.11/site-packages/pyface/ui/qt/gui.py", line 190, in _dispatch'
[2024-06-12T15:33:10.220Z] b'    self._callable(*self._args, **self._kw)'
[2024-06-12T15:33:10.220Z] b'  File "/home/builder/.edm/envs/build-env/lib/python3.11/site-packages/pyface/application.py", line 220, in exit'
[2024-06-12T15:33:10.220Z] b'    self._exit()'
[2024-06-12T15:33:10.220Z] b'  File "/home/builder/.edm/envs/build-env/lib/python3.11/site-packages/pyface/gui_application.py", line 236, in _exit'
[2024-06-12T15:33:10.220Z] b'    self.gui.stop_event_loop()'
[2024-06-12T15:33:10.220Z] b'    ^^^^^^^^^^^^^^^^^^^^^^^^'
[2024-06-12T15:33:10.220Z] b"AttributeError: '_Undefined' object has no attribute 'stop_event_loop'"
[2024-06-12T15:33:10.220Z] b'FAIL'
[2024-06-12T15:33:10.220Z] b''
[2024-06-12T15:33:10.220Z] b'======================================================================'
[2024-06-12T15:33:10.220Z] b'FAIL: test2_test_pyside6 (__main__.Tests.test2_test_pyside6)'
[2024-06-12T15:33:10.220Z] b'----------------------------------------------------------------------'
[2024-06-12T15:33:10.220Z] b'Traceback (most recent call last):'
[2024-06-12T15:33:10.220Z] b'  File "/home/builder/workspace/Buildsystem/test/new/cp311/pyface-8.0.0/tests/run.py", line 27, in test2_test_pyside6'
[2024-06-12T15:33:10.220Z] b'    self.runpy(args, env=env)'
[2024-06-12T15:33:10.220Z] b'  File "/home/builder/workspace/Buildsystem/test/new/cp311/entest_utils.py", line 469, in runpy'
[2024-06-12T15:33:10.220Z] b'    self.run_test_function('
[2024-06-12T15:33:10.220Z] b'  File "/home/builder/workspace/Buildsystem/test/new/cp311/entest_utils.py", line 546, in run_test_function'
[2024-06-12T15:33:10.220Z] b'    self.fail("unexpected test exit code {!r}".format(ret))'
[2024-06-12T15:33:10.220Z] b'AssertionError: unexpected test exit code -15'
[2024-06-12T15:33:10.220Z] b''
[2024-06-12T15:33:10.220Z] b'----------------------------------------------------------------------'
[2024-06-12T15:33:10.220Z] b'Ran 18 tests in 1180.661s'
[2024-06-12T15:33:10.220Z] b''
[2024-06-12T15:33:10.220Z] b'FAILED (failures=1, skipped=11)'
[2024-06-12T15:33:10.220Z] b'================================================='
[2024-06-12T15:33:10.220Z] b"ENTEST STOP -- TOTAL FAILURES: 1 ('pyface-8.0.0')"
```

--- a/pyface/tasks/tests/test_tasks_application.py
+++ b/pyface/tasks/tests/test_tasks_application.py
@@ -134,10 +134,14 @@ class TestApplication(unittest.TestCase, GuiTestAssistant):
         app = TasksApplication()
         self.connect_listeners(app)
         window = ApplicationWindow()
-        app.observe(lambda _: app.add_window(window), "started")
+
+        def on_started(event):
+            app.add_window(window)
+            self.gui.invoke_after(100, app.exit)
+
+        app.observe(on_started, "started")
 
         with self.assertMultiTraitChanges([app], EVENTS, []):
-            self.gui.invoke_after(1000, app.exit)
             result = app.run()
 
         self.assertTrue(result)

--- a/pyface/tests/test_gui_application.py
+++ b/pyface/tests/test_gui_application.py
@@ -167,8 +167,12 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         window = ApplicationWindow()
         app.observe(lambda _: app.add_window(window), "started")
 
+        def on_started(event):
+            self.gui.invoke_after(100, app.exit)
+
+        app.observe(on_started, "started")
+
         with self.assertMultiTraitChanges([app], EVENTS, []):
-            self.gui.invoke_after(1000, app.exit)
             result = app.run()
 
         self.assertTrue(result)
@@ -180,9 +184,14 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         app = TestingApp(exit_prepared_error=True)
         self.connect_listeners(app)
 
-        with self.assertMultiTraitChanges([app], EVENTS, []):
-            self.gui.invoke_after(1000, app.exit)
-            result = app.run()
+        def on_started(event):
+            self.gui.invoke_after(100, app.exit)
+
+        app.observe(on_started, "started")
+
+        with self.assertLogs("pyface.application", "WARNING") as logs_cm:
+            with self.assertMultiTraitChanges([app], EVENTS, []):
+                result = app.run()
 
         self.assertTrue(result)
         self.assertFalse(app.exit_vetoed)
@@ -190,6 +199,10 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         event_order = [event.event_type for event in self.application_events]
         self.assertEqual(event_order, EVENTS)
         self.assertEqual(app.windows, [])
+
+        self.assertEqual(len(logs_cm.output), 1)
+        [log_output] = logs_cm.output
+        self.assertIn("Error preparing for application exit", log_output)
 
     def test_veto_exit(self):
         app = TestingApp(veto_exit=True)
@@ -211,8 +224,12 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         app = TestingApp(veto_open_window=True)
         self.connect_listeners(app)
 
-        with self.assertMultiTraitChanges([app], EVENTS, []):
+        def on_started(event):
             self.gui.invoke_after(1000, app.exit)
+
+        app.observe(on_started, "started")
+
+        with self.assertMultiTraitChanges([app], EVENTS, []):
             result = app.run()
 
         self.assertTrue(result)
@@ -227,9 +244,13 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         app = TestingApp(veto_close_window=True)
         self.connect_listeners(app)
 
-        with self.assertMultiTraitChanges([app], EVENTS, []):
+        def on_started(event):
             self.gui.invoke_after(1000, app.exit)
             self.gui.invoke_after(2000, app.exit, force=True)
+
+        app.observe(on_started, "started")
+
+        with self.assertMultiTraitChanges([app], EVENTS, []):
             result = app.run()
 
         self.assertTrue(result)
@@ -243,8 +264,12 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         app = TestingApp(do_exit=True, force_exit=True, veto_exit=True)
         self.connect_listeners(app)
 
-        with self.assertMultiTraitChanges([app], EVENTS, []):
+        def on_started(event):
             self.gui.invoke_after(100, app.exit, True)
+
+        app.observe(on_started, "started")
+
+        with self.assertMultiTraitChanges([app], EVENTS, []):
             result = app.run()
 
         self.assertTrue(result)
@@ -258,8 +283,12 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         app = TestingApp(do_exit=True, force_exit=True, veto_close_window=True)
         self.connect_listeners(app)
 
+        def on_started(event):
+            self.gui.invoke_after(100, app.exit, True)
+
+        app.observe(on_started, "started")
+
         with self.assertMultiTraitChanges([app], EVENTS, []):
-            self.gui.invoke_after(1000, app.exit, True)
             result = app.run()
 
         self.assertTrue(result)
@@ -285,8 +314,12 @@ class TestGUIApplication(unittest.TestCase, GuiTestAssistant):
         app = TestingApp(stop_cleanly=False)
         self.connect_listeners(app)
 
+        def on_started(event):
+            self.gui.invoke_after(100, app.exit)
+
+        app.observe(on_started, "started")
+
         with self.assertMultiTraitChanges([app], EVENTS[:-1], EVENTS[-1:]):
-            self.gui.invoke_after(1000, app.exit, True)
             result = app.run()
 
         self.assertFalse(result)

